dimensiondata/src/main/java/org/jclouds/dimensiondata/cloudcontrol/compute/functions/OsImageToImage.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jclouds.dimensiondata.cloudcontrol.compute.functions;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Function;
+import com.google.common.base.Splitter;
+import com.google.common.base.Supplier;
+import com.google.common.collect.FluentIterable;
+import com.google.inject.Inject;
+import org.jclouds.collect.Memoized;
+import org.jclouds.compute.domain.Image;
+import org.jclouds.compute.domain.ImageBuilder;
+import org.jclouds.compute.domain.OperatingSystem;
+import org.jclouds.compute.domain.OsFamily;
+import org.jclouds.dimensiondata.cloudcontrol.domain.OsImage;
+import org.jclouds.domain.Location;
+import org.jclouds.location.predicates.LocationPredicates;
+
+import javax.inject.Singleton;
+import java.util.Set;
+
+import static com.google.common.collect.Iterables.get;
+
+@Singleton
+public class OsImageToImage implements Function<OsImage, Image> {
+
+   private final Supplier<Set<Location>> locations;
+   private final Function<String, OsFamily> imageDescriptionToOsFamily;
+
+   @Inject
+   OsImageToImage(@Memoized final Supplier<Set<Location>> locations,
+         Function<String, OsFamily> imageDescriptionToOsFamily) {
+      this.locations = locations;
+      this.imageDescriptionToOsFamily = imageDescriptionToOsFamily;
+   }
+
+   @Override
+   public Image apply(OsImage input) {
+
+      OsFamily osFamily = imageDescriptionToOsFamily.apply(input.description());
+      String osVersion = parseVersion(input.description());
+
+      OperatingSystem os = OperatingSystem.builder().name(input.name()).description(input.description())
+            .family(osFamily).version(osVersion).is64Bit(is64bit(input)).build();
+
+      return new ImageBuilder().id(input.id()).name(input.name()).description(input.description())
```

Comment:
When configuring the image IDs, think about the "id" and the "providerId" field this way:

* **providerId** is the *real id* of the image in the provider (I guess `input.id()`), so users can retrieve the image using the provider-specific APIs.
* **id** is the *portable id jclouds will use to uniquelly identify the image*. This is slightly different from the previous one. For example, in some providers, images in different regions may have the same id (centos-6, ami-0001, etc), but jcouds needs to provide unique IDs for each, so we encode the region in the id (something like: "region/id").

I don't know if it is the case in DimensionData, but please consider this when configuring the id and providerId fields. If the ID is unique, then just assign the same value to both fields :)

Reply:
@nacx thanks for the background information. In DimensionData each image is associated with a single datacenter (zone) and will the ID will be unique within a Region. Therefore I will keep the `input.id()` as is.

Reply:
Is it possible that that unique ID within that region collisions with a unique id within another region?

Reply:
I mean, you should think about it this way: calling `ComputeService.getImage(id)` with this ID will succeed? Or should the ID carry/encode additional info?

Reply:
Perfect! Thanks for clarifying. Then, mind just changing the code to `ImageBuilder().ids(input.id())` (note the plural there)? This will set both the id and providerId fields.